### ISSUE TITLE
want to inject a specific InputDecorator into TextField and TextFormField

### ISSUE THREAD

**Issue by SteveAlexander** [flutter/flutter#19775]
I want to create a Text form field that has an animated hint. When the field is empty, the hint text is in its normal place, where the field's input text would go. When there is input text, the hint is at a smaller font size, and sits above the input text.

A colleague made this work by creating a Widget that contains two TextFields, one for the input and one for the text, and doing animations based on what's going on with the input TextField.

This is fine, but looking at how TextFields and TextFormFields work, I feel like this ought to be the responsibility of an InputDecorator, taking a standard set of InputDecoration, and choosing to display it in a slightly different way than the regular InputDecorator.

However, it looks hard to replace the InputDecorator that a TextField or TextFormField uses. It's hard-coded to be an InputDecorator deep within the TextField's `build` method, and to make a TextFormField that uses a different InputDecorator, I imagine I'd need to reimplement / copy-paste a lot of stuff, so I could change the InputDecorator implementation.

tl;dr please make a way to inject a specific InputDecorator into TextField and TextFormField.

Flutter (Channel dev, v0.5.7, on Mac OS X 10.13.5 17F77, locale en-GB)

**Comment by HansMuller** [comment]
> This is fine, but looking at how TextFields and TextFormFields work, I feel like this ought to be the responsibility of an InputDecorator, taking a standard set of InputDecoration, and choosing to display it in a slightly different way than the regular InputDecorator.
> 
> However, it looks hard to replace the InputDecorator that a TextField or TextFormField uses. It's hard-coded to be an InputDecorator deep within the TextField's build method, and to make a TextFormField that uses a different InputDecorator, I imagine I'd need to reimplement / copy-paste a lot of stuff, so I could change the InputDecorator implementation.

Sorry, I'm not completely understanding this. Is it that you want to specify InputDecorator subclass for the TextField's `decoration` parameter but you're finding it difficult to build an InputDecorator subclass that changes the input decorator's default layout or behavior?

**Comment by SteveAlexander** [comment]
Thanks @HansMuller. I'll have a go at doing a specialized InputDecorator for the situation I described, and report back.

I think I understand what you mean by the "more Fluttery" approach too — something like being able to give a TextField scaffold-type thing that orchestrates the layout and behaviour of the various parts that make up a text field — editable text, hint, icon, prefix/suffix, other decorations.

**Comment by zoechi** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. We are therefore reluctantly going to close this bug for now. Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!
Thanks for your contribution.
